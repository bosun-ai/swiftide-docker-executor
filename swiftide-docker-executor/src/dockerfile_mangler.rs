/// Adds copy statements to the Dockerfile to copy the built binary into the image.
use std::path::Path;

use tokio::fs::read_to_string;

use crate::MangleError;

pub struct MangledDockerfile {
    pub content: String,
}

pub async fn mangle(path: &Path) -> Result<MangledDockerfile, MangleError> {
    tracing::warn!("Mangling Dockerfile at {:?}", path);

    let this_crate = env!("CARGO_PKG_VERSION");
    let mut content = read_to_string(path)
        .await
        .map_err(MangleError::DockerfileReadError)?;

    let image_name = format!("bosunai/swiftide-docker-service:{this_crate}");
    // Remove existing CMD or ENTRYPOINT instructions
    content = content
        .lines()
        .filter(|line| {
            !line.trim_start().to_lowercase().starts_with("cmd")
                && !line.trim_start().to_lowercase().starts_with("entrypoint")
        })
        .collect::<Vec<&str>>()
        .join("\n");

    // Find the position to insert the copy lines
    let mut lines = content.lines().collect::<Vec<_>>();
    let insert_pos = lines
        .iter()
        .enumerate()
        .filter(|(_, line)| line.trim_start().to_lowercase().starts_with("from"))
        .last()
        .map(|(idx, _)| idx)
        .ok_or(MangleError::InvalidDockerfile)?
        + 1;

    // Copy swiftide-docker-service, rg, and fd into the image
    let copy_lines = ["swiftide-docker-service", "rg", "fd"]
        .iter()
        .map(|binary| format!("COPY --from={image_name} /usr/bin/{binary} /usr/local/bin/{binary}"))
        .collect::<Vec<String>>()
        .join("\n");

    lines.insert(insert_pos, &copy_lines);

    let new_dockerfile = lines.join("\n");
    tracing::debug!(
        original = content,
        mangled = new_dockerfile,
        "Mangled dockerfile"
    );
    Ok(MangledDockerfile {
        content: new_dockerfile,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs::File;
    use std::io::Write;
    use tempfile::tempdir;

    // Normalizes the the dockerfile versions for consistent tests
    macro_rules! assert_snapshot {
        ($content:expr) => {{
            insta::with_settings!({filters => vec![
                (r"\b\d{1,2}\.\d{1,2}\.\d{1,2}", "[CARGO_PKG_VERSION]")
            ]}, {
                insta::assert_snapshot!($content);
            });
        }}
    }

    #[tokio::test]
    async fn test_mangle() {
        let dir = tempdir().unwrap();
        let file_path = dir.path().join("Dockerfile");
        let mut file = File::create(&file_path).unwrap();
        writeln!(file, "FROM alpine").unwrap();

        let result = mangle(&file_path).await.unwrap();

        assert!(
            result
                .content
                .contains("COPY --from=bosunai/swiftide-docker-service:"),
            "action {}",
            result.content
        );
        assert!(result
            .content
            .contains("/usr/local/bin/swiftide-docker-service"));
        assert!(result.content.contains("/usr/local/bin/rg"));
        assert!(result.content.contains("/usr/local/bin/fd"));
        assert_snapshot!(result.content)
    }

    #[tokio::test]
    async fn test_remove_cmd_entrypoint() {
        let dir = tempdir().unwrap();
        let file_path = dir.path().join("Dockerfile");
        let mut file = File::create(&file_path).unwrap();
        writeln!(
            file,
            "FROM alpine\nCMD [\"echo\", \"Hello World\"]\nENTRYPOINT [\"/bin/sh\"]"
        )
        .unwrap();

        let result = mangle(&file_path).await.unwrap();

        assert!(
            !result.content.contains("CMD [\"echo\", \"Hello World\"]"),
            "actual {}",
            result.content
        );
        assert!(!result.content.contains("ENTRYPOINT [\"/bin/sh\"]"));
        assert_snapshot!(result.content)
    }

    #[tokio::test]
    async fn test_remove_cmd_entrypoint_case_insensitive() {
        let dir = tempdir().unwrap();
        let file_path = dir.path().join("Dockerfile");
        let mut file = File::create(&file_path).unwrap();
        writeln!(
            file,
            "from alpine\ncmd [\"echo\", \"Hello World\"]\neNtryPoint [\"/bin/sh\"]"
        )
        .unwrap();

        let result = mangle(&file_path).await.unwrap();

        assert!(!result.content.contains("CMD [\"echo\", \"Hello World\"]"));
        assert!(!result.content.contains("ENTRYPOINT [\"/bin/sh\"]"));
        assert_snapshot!(result.content)
    }

    #[tokio::test]
    async fn test_mangle_with_multiple_forms_and_other_lines() {
        let dir = tempdir().unwrap();
        let file_path = dir.path().join("Dockerfile");
        let mut file = File::create(&file_path).unwrap();
        writeln!(
            file,
            "FROM alpine\nFROM ubuntu\nRUN echo hello world\nCMD [\"true\"]"
        )
        .unwrap();

        let result = mangle(&file_path).await.unwrap();

        assert_snapshot!(result.content)
    }
}
