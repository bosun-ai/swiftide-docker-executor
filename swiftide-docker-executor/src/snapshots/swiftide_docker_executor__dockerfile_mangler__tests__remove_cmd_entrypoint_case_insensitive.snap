---
source: swiftide-docker-executor/src/dockerfile_mangler.rs
expression: result.content
---
from alpine
COPY --from=bosun-ai/swiftide-docker-service:[CARGO_PKG_VERSION] /usr/bin/swiftide-docker-service /usr/local/bin/swiftide-docker-service
COPY --from=bosun-ai/swiftide-docker-service:[CARGO_PKG_VERSION] /usr/bin/rg /usr/local/bin/rg
COPY --from=bosun-ai/swiftide-docker-service:[CARGO_PKG_VERSION] /usr/bin/fd /usr/local/bin/fd
